synapse/handlers/device.py

```diff
@@ -671,12 +671,12 @@ async def delete_all_devices_for_user(
             except_device_id: optional device id which should not be deleted
         """
         device_map = await self.store.get_devices_by_user(user_id)
-        device_ids = list(device_map)
+        device_ids = set(device_map)
         if except_device_id is not None:
-            device_ids = [d for d in device_ids if d != except_device_id]
+            device_ids.discard(except_device_id)
         await self.delete_devices(user_id, device_ids)
```

Comment:
```suggestion
        device_map = await self.store.get_devices_by_user(user_id)

        if except_device_id is not None:
            device_map.pop(except_device_id, None)

        user_device_ids = device_map.keys()
        await self.delete_devices(user_id, user_device_ids)
```

to avoid the extra allocation.

Reply:
Hmm, true. Always slight wary about mutating things we get from the store in case they're cached, but I guess we have linters to catch that nowadays